File: Individual_Assignment/MatrixMultiplication/code/c/app/bench.c
#include "matrix.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <math.h>

static double now_sec(void){
#if defined(_POSIX_TIMERS)
    struct timespec ts; clock_gettime(CLOCK_MONOTONIC, &ts);
    return (double)ts.tv_sec + 1e-9*ts.tv_nsec;
#else
    return (double)clock()/CLOCKS_PER_SEC;
#endif
}

static Order parse_order(const char* s){
    if(!s) return ORDER_IKJ;
    if(strcmp(s,"ijk")==0) return ORDER_IJK;
    if(strcmp(s,"ikj")==0) return ORDER_IKJ;
    if(strcmp(s,"jik")==0) return ORDER_JIK;
    return ORDER_IKJ;
}

int main(int argc, char** argv){
    size_t n = 512;
    int runs = 5;
    unsigned int seed = 42;
    const char* order_s = "ikj";

    for (int i=1; i<argc; ++i){
        if (!strcmp(argv[i],"-n") && i+1<argc) n = (size_t) atoll(argv[++i]);
        else if (!strcmp(argv[i],"-r") && i+1<argc) runs = atoi(argv[++i]);
        else if (!strcmp(argv[i],"-s") && i+1<argc) seed = (unsigned) atoi(argv[++i]);
        else if (!strcmp(argv[i],"-o") && i+1<argc) order_s = argv[++i];
    }
    Order order = parse_order(order_s);

    Matrix A = mat_alloc(n), B = mat_alloc(n), C = mat_alloc(n);
    mat_fill_random(A, seed);
    mat_fill_random(B, seed+1);

    double *times = malloc((size_t)runs * sizeof(double));

    for (int r=0; r<runs; ++r){
        mat_zero(C);
        double t0 = now_sec();
        mat_mul(A, B, C, order);
        double t1 = now_sec();
        times[r] = t1 - t0;
        fprintf(stderr, "run %d/%d: %.6f s\n", r+1, runs, times[r]);
        if (C.data[0] == -1.0) puts("!");
    }

    double sum=0.0; for (int i=0;i<runs;++i) sum += times[i];
    double mean = sum / runs;
    double var=0.0; for (int i=0;i<runs;++i){ double d=times[i]-mean; var += d*d; }
    double stdev = runs>1 ? sqrt(var/(runs-1)) : 0.0;

    printf("{\"lang\":\"C\",\"n\":%zu,\"runs\":%d,\"order\":\"%s\","
           "\"mean_sec\":%.6f,\"stdev_sec\":%.6f}\n",
           n, runs, order_s, mean, stdev);

    free(times);
    mat_free(&A); mat_free(&B); mat_free(&C);
    return 0;
}

File: Individual_Assignment/MatrixMultiplication/code/python/bench.py
import argparse, time, statistics, json, tracemalloc
from src.matrix import rand_matrix, matmul

def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("-n", type=int, default=1024)
    ap.add_argument("-r", "--runs", type=int, default=5)
    ap.add_argument("-s", "--seed", type=int, default=42)
    ap.add_argument("-o", "--order", choices=["ijk","ikj","jik"], default="ikj")
    ap.add_argument("--mem", action="store_true", help="medir memoria (tracemalloc)")
    args = ap.parse_args()

    A = rand_matrix(args.n, args.seed)
    B = rand_matrix(args.n, args.seed+1)

    times = []
    peak = None
    if args.mem: tracemalloc.start()

    for _ in range(args.runs):
        t0 = time.perf_counter()
        C = matmul(A, B, args.order)
        t1 = time.perf_counter()
        times.append(t1 - t0)
        if C[0][0] == 1.23456789: print("!")  # evita DCE

    if args.mem:
        _, peak = tracemalloc.get_traced_memory()
        tracemalloc.stop()

    out = {
        "lang": "Python",
        "n": args.n,
        "runs": args.runs,
        "order": args.order,
        "mean_sec": statistics.fmean(times),
        "stdev_sec": statistics.pstdev(times) if len(times) > 1 else 0.0
    }
    if peak is not None: out["peak_bytes"] = int(peak)
    print(json.dumps(out))

if __name__ == "__main__":
    main()
